lib/archethic/transaction_chain/transaction/validation_stamp/ledger_operations.ex

```diff
@@ -413,22 +422,56 @@ defmodule Archethic.TransactionChain.Transaction.ValidationStamp.LedgerOperation
     |> Enum.group_by(& &1.type)
     |> Enum.flat_map(fn
       {:UCO, inputs} ->
-        include_uco? = uco_to_spend > 0
-        if include_uco? or consolidate_inputs?(inputs), do: inputs, else: []
+        get_uco_to_consume(inputs, uco_to_spend, uco_balance)
 
       {{:token, token_address, token_id}, inputs} ->
-        token_used? = Map.has_key?(tokens_to_spend, {token_address, token_id})
-        if token_used? or consolidate_inputs?(inputs), do: inputs, else: []
+        key = {token_address, token_id}
+        get_token_to_consume(inputs, key, tokens_to_spend, tokens_balance)
 
       {:state, [state_utxo = %UnspentOutput{encoded_payload: previous_state}]} ->
         if encoded_state != nil && previous_state != encoded_state, do: [state_utxo], else: []
 
       {:call, inputs} ->
-        get_contract_call_input(inputs, contract_context)
+        get_call_to_consume(inputs, contract_context)
     end)
   end
 
-  defp get_contract_call_input(inputs, %ContractContext{trigger: {:transaction, address, _}}) do
+  defp get_uco_to_consume(inputs, uco_to_spend, uco_balance) do
```

Comment:
Suggestion for readability to avoid nested blocks of condition:

```elixir
defp get_uco_to_consume(inputs, uco_to_spend, uco_balance) when uco_to_spend > 0 do
  optimize_inputs_to_consume(inputs, uco_to_spend, uco_balance)
end

defp get_uco_to_consume(inputs, _uco_to_spend, _uco_balance) when length(inputs) > 0, do: inputs
defp get_uco_to_consume(_inputs, _uco_to_spend, _uco_balance), do: []

defp get_token_to_consume(inputs, key, tokens_to_spend, tokens_balance) when is_map_key(tokens_to_spend, key) do
  amount_to_spend = Map.get(tokens_to_spend, key)
  token_balance = Map.get(tokens_balance, key)
  optimize_inputs_to_consume(inputs, amount_to_spend, token_balance)
end

defp get_token_to_consume(inputs, _key, _tokens_to_spend, _tokens_balance) when length(inputs) > 0, do: inputs
defp get_token_to_consume(_inputs, _key, _tokens_to_spend, _tokens_balance), do: []
```

Reply:
Done, thank's